### PR TITLE
Add Awaited<T> to core library

### PR DESCRIPTION
**Pull request by rbuckton**
This adds `Awaited<T>` as a mechanism to unwrap a type to its "awaited" type:

```ts
declare type Awaited<T> =
    T extends { then(onfulfilled: (value: infer U) => any): any; } ? U :
    T extends { then(...args: any[]): any; } ? never :
    T;
```

Unlike the `await` expression, `Awaited<T>` does *not* recursively unwrap `T` as we do not support circular references in type aliases. It turns out recursive unwrap of `T` is mostly unnecessary if we modify the definition of `Promise` and `PromiseLike` to return a `Promise<Awaited<TResult>>`:

```ts
interface Promise<T> {
    then<TResult1 = T, TResult2 = never>(
        onfulfilled?: ((value: Awaited<T>) => TResult1) | null, 
        onrejected?: ((reason: any) => TResult2) | null
    ): Promise<Awaited<TResult1 | TResult2>>;
}
```

This effectively flattens `Promise<Promise<number>>` as well as other nested definitions of `Promise`:

```ts
type T1 = Awaited<number>; // number
type T2 = Awaited<Promise<number>>; // number
type T3 = Awaited<Promise<Promise<number>>>; // number
```

The `Awaited<T>` type is a conditional type that provides the following conditions:

- If `T` has a `then()` method with a callback, infer the callback's first argument as `U` and use that as the result type.
- If `T` has a `then()` method without a callback, this is a non-promise "thenable" that can never be resolved, so the result type is `never`.
- Otherwise, the result type is `T`.

The benefit of this approach is more effective type inference from the `onfulfilled` callback passed to the `then` method of a `Promise`:

```ts
const p: Promise<number> = ...;
const x = p.then(value => {
  if (value > 1) return Promise.resolve("a");
  return false;
});
// `x` has the type `Promise<"a" | false>`
```

### CODE REVIEW

**Comment by rbuckton** [comment]
This only solves the "Incorrect return types or errors with complex Promise chains" issue in https://github.com/Microsoft/TypeScript/pull/17077.

This does not solve the "Incorrect eager resolution of the "awaited type" for a type variable" issue yet.

**Comment by ajafff** [comment]
There's a difference to `await` if the on fulfilled callback has no parameter. `await` unwraps the promise to `void` while `Awaited` resolves to `never` AFAICT

File: src/lib/es5.d.ts [hunk 1]
```diff
@@ -1273,7 +1273,20 @@ declare type PropertyDecorator = (target: Object, propertyKey: string | symbol)
...
-declare type PromiseConstructorLike = new <T>(executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void) => PromiseLike<T>;
+/** An object type that can be definitely be awaited. Do not inherit from this type. */
+declare type Awaitable<T> = { then(onfulfilled: (value: T) => any): any; };
+
+/** An object type that may not be awaitable. Do not inherit from this type. */
+declare type NonAwaitable = { then(...args: any[]): any; };
+
+/** Gets the type resulting from awaiting `T`. This does **not** recursively unwrap nested promises. */
+declare type Awaited<T> =
+    T extends Awaitable<Awaitable<infer U>> ? U :
```
Comment (weswigham) [inline src/lib/es5.d.ts]: Do we figure that two levels deep is enough for most uses, and if anyone complains, we can always add more later?

File: src/lib/es5.d.ts [hunk 2]
```diff
@@ -1273,7 +1273,20 @@ declare type PropertyDecorator = (target: Object, propertyKey: string | symbol)
 declare type MethodDecorator = <T>(target: Object, propertyKey: string | symbol, descriptor: TypedPropertyDescriptor<T>) => TypedPropertyDescriptor<T> | void;
 declare type ParameterDecorator = (target: Object, propertyKey: string | symbol, parameterIndex: number) => void;
 
-declare type PromiseConstructorLike = new <T>(executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void) => PromiseLike<T>;
+/** An object type that can be definitely be awaited. Do not inherit from this type. */
+declare type Awaitable<T> = { then(onfulfilled: (value: T) => any): any; };
+
+/** An object type that may not be awaitable. Do not inherit from this type. */
```
Comment (weswigham) [inline src/lib/es5.d.ts]: If this type isn't supposed to be extended/used, why not just inline the object type inside `Awaited`, below?

**Review by weswigham — APPROVED** [review]
It definitely seems better than the union with `PromiseLike` most positions took before from before, and we can definitely improve it in the future if we can.

We should audit how this affects definitelytyped, though.

**Comment by Igorbek** [comment]
Nice! This is what @tycho01 [suggested](https://github.com/Microsoft/TypeScript/pull/17077#issuecomment-315626274) by leveraging #6606 but now with conditional types.

File: src/lib/es5.d.ts [hunk 3]
```diff
@@ -1273,7 +1273,20 @@ declare type PropertyDecorator = (target: Object, propertyKey: string | symbol)
 declare type MethodDecorator = <T>(target: Object, propertyKey: string | symbol, descriptor: TypedPropertyDescriptor<T>) => TypedPropertyDescriptor<T> | void;
 declare type ParameterDecorator = (target: Object, propertyKey: string | symbol, parameterIndex: number) => void;
 
-declare type PromiseConstructorLike = new <T>(executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void) => PromiseLike<T>;
+/** An object type that can be definitely be awaited. Do not inherit from this type. */
```
Comment (DanielRosenwasser) [inline src/lib/es5.d.ts]: "be definitely be"

File: src/lib/es5.d.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/lib/es5.d.ts]: Duh, two levels of conditional type inference should be enough for anyone™

Comment (DanielRosenwasser) [reply src/lib/es5.d.ts]: I guess if we currently don't do better than two levels, this will do for now.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: src/lib/es5.d.ts (hunk 1 above)
Comment (rbuckton) [reply src/lib/es5.d.ts]: If you are using the `Promise` definition from our libs then you won't need anything deeper. Once this is shipping we can modify the definitions in DT to use it as well. Per my examples above: `type T3 = Awaited<Promise<Promise<number>>>; // number` for any level of nesting using either `Promise` or `PromiseLike`.

File: src/lib/es5.d.ts (hunk 2 above)
Comment (rbuckton) [reply src/lib/es5.d.ts]: I've considered that, but this is more readable when getting Quick Info on `Awaited<T>`. I don't even need `Awaitable<T>` if I inline it in the 2-3 places its used but that again reduces readability.

**Comment by rbuckton** [comment]
@ajafff I'll see if there's anything worth changing here. Despite the naming `Awaited<T>` isn't the same as an `await` expression as its currently only used in the definition of `Promise` (and `PromiseLike`).

File: src/lib/es5.d.ts (hunk 1 above)
Comment (rbuckton) [reply src/lib/es5.d.ts]: We actually only need:
```ts
declare type Awaited<T> =
    T extends Awaitable<infer U> ? U :
    T extends { then(...args: any[]): any; } ? never :
    T;
```

I seemed to have forgotten to stage an additional change to es5.d.ts.

File: src/lib/es5.d.ts (hunk 2 above)
Comment (rbuckton) [reply src/lib/es5.d.ts]: I have removed `NonAwaitable`.

File: src/lib/es5.d.ts (hunk 3 above)
Comment (rbuckton) [reply src/lib/es5.d.ts]: Fixed

**Comment by rbuckton** [comment]
> There's a difference to `await` if the on fulfilled callback has no parameter. `await` unwraps the promise to `void` while `Awaited` resolves to `never` AFAICT

@ajafff actually, if the `onfulfilled` callback doesn't take a parameter then we report an error:

```ts
async function foo() {
  let y: { then(): void; };
  let x = await y; // Type of 'await' operand must either be a valid promise or must not contain a callable 'then' member.
}
```

The best approximation of an error we can provide here is to resolve the type to `never`, which is what we are doing.

File: src/lib/es5.d.ts (hunk 2 above)
Comment (rbuckton) [reply src/lib/es5.d.ts]: I decided to also remove `Awaitable<T>` and inline the definition into `Awaited<T>`. The two other places where it was referenced were fine as `PromiseLike<T>`.

**Comment by sandersn** [comment]
Can you run this on DefinitelyTyped before checking in? Any change to inference is likely to disturb things there and it would be nice to know how big the break is.

**Review by mhegazy — APPROVED** [review]

**Comment by mhegazy** [comment]
Please also run the RWC tests before submitting and ensure we have not introduced any regressions there either.

**Comment by mhegazy** [comment]
and user tests.

**Comment by rbuckton** [comment]
@mhegazy, I plan to run this against RWC and DefinitelyTyped today. If all goes well, should we discuss this in the design meeting before taking the change?

**Comment by mhegazy** [comment]
> If all goes well, should we discuss this in the design meeting before taking the change?

sure. We should discuss the `while` types as well to enable recursive unwrapping. Just had a chat with @ahejlsberg about a few minutes ago.

**Comment by rbuckton** [comment]
@mhegazy There are a few failing RWC tests after this that may actually be more correct. Most of them seem to be related to assignability, i.e.: `Type 'Awaited<T> is not assignable to type 'T'`. I will dig into these more tomorrow.

**Comment by rbuckton** [comment]
An example of one of the RWC errors is this:

```
Type 'Promise<Awaited<T>>' is not assignable to type 'Promise<T>'.
  Type 'Awaited<T>' is not assignable to type 'T'.
    Type 'U | (T extends { then(...args: any[]): any; } ? never : T)' is not assignable to type 'T'.
      Type 'U' is not assignable to type 'T'.
```

I'd expect `Promise<Awaited<T>>` to be synonymous with `Promise<T>`, except that we see that both are type references to `Promise` and `Awaited<T>` and `T` are not comparable. If we had attempted a structural comparison of `Promise` we still would fail when instantiating the `onfulfilled` callback to `then` as we would end up comparing `Awaited<Awaited<T>>` to `Awaited<T>`, which still aren't comparable. 

@ahejlsberg, any suggestions? Is there something I can change in my definition to address this issue, or is there a change we can make in the checker that makes sense? Is this something that `while` types could address?

**Comment by tycho01** [comment]
I got it to recursively unwrap under the present definition of `Promise` as follows:
```ts
type Awaited<T> = {
  '0': T;
  '1': Awaited<T extends { then(onfulfilled: (value: infer U) => any): any; } ? U : 'wat'>;
}[T extends { then(onfulfilled: (value: any) => any): any; } ? '1' : '0'];

let one: 1;
one = null! as Awaited<1>>; // ok
one = null! as Awaited<Promise<1>>>; // ok
one = null! as Awaited<Promise<Promise<1>>>>; // ok
```
I'll admit it's clunky and silly -- I just used an object for the conditional to enable type recursion. But it works on master already, fwiw.

**Comment by weswigham** [comment]
@tycho01 @ahejlsberg  on `master` I see this example:
```ts
type Awaited<T> = {
  '0': T;
  '1': Awaited<T extends { then(onfulfilled: (value: infer U) => any): any; } ? U : never>;
}[T extends { then(onfulfilled: (value: any) => any): any; } ? '1' : '0'];


function foo<T>(x: T) {
  let one: T;
  let n1: Awaited<T>;
  let n2: Awaited<Awaited<T>>;
  let n3: Awaited<Awaited<Awaited<T>>>;
  one = n1;
  one = n2;
  one = n3;
}
```
spin forever. Looks like an infinite loop in `instnatiateType`, during `computeBaseConstraint`.

**Comment by rbuckton** [comment]
@weswigham, I think that is the same issue as in #21611

**Comment by tycho01** [comment]
Sorry, just remembered I'd yet to test for the original core issue, [`1|Promise<1>`](https://github.com/tycho01/typical/commit/811acec846581027ccec0b990283dd7d02857e5f). I've yet to make that work.
Also, I mistakenly said master, while I was on `2.8.0-dev.20180210`.

**Comment by tycho01** [comment]
I played around a bit more to get the union cases to work properly.

Conditional types already distributed over union types properly, so this came down to finding a way to make recursive types based on this, i.e. somehow wrapping the recursion so it'd allow it, using some useless no-op wrapper delaying execution to trick it.

I found a silly approach that appears to work: `{ '1': ... }[T extends number ? '1' : '1']`. Here `T` is the generic we use to delay unwrapping, while `number` is some arbitrary condition -- the point is this wrapper is there during the recursion check, and vanishes once `T` is plugged in.

Using that to unwrap promises or to flatten nested array types:
```ts
// unwrap a Promise to obtain its return value
export type Awaited<T> = {
  '1': T extends { then(onfulfilled: (value: infer U) => any): any; } ? Awaited<U> : T;
}[T extends number ? '1' : '1'];

// Awaited<Promise<1 | Promise<2>>> // -> 1 | 2

// flatten a structure of nested tuples/arrays into a flat array type
export type Flatten<T> = {
  '1': Flatten<T extends Array<infer U> ? U : T>;
}[T extends number ? '1' : '1'];

// Flatten<Array<1 | Array<2>>> // -> 1 | 2
```

**Comment by ericanderson** [comment]
@tycho01 you rock my friend

**Comment by tycho01** [comment]
Sorry, so far I'd tested through the compiler API, but through `tsc` I ran into a recursion related issue on that `Flatten`, now posted at #22018.

**Comment by mhegazy** [comment]
Does not look like this is something we can proceed on at the time being.closing the PR, but leaving the branch for further investigations.

**Pull request closed without merging**